assessments/projects/keycloak/self-assessment.md

```diff
@@ -0,0 +1,733 @@
+        *   **_If an access token or refresh token is compromised, the first thing you should do is go to the admin console and push a not-before revocation policy to all applications. This will enforce that any tokens issued prior to that date are now invalid. Pushing a new not-before policy will also ensure that an application will be forced to download new public keys from Keycloak, hence it is also useful for the case, when you think that realm signing key was compromised. More info in the [keys chapter](https://www.keycloak.org/docs/latest/server_admin/#realm_keys)._**
+        *   **_You can also disable specific applications, clients, and users if you feel that any one of those entities is completely compromised.”_**
+    *   _Eg. _[Compromised Authorization Code](https://www.keycloak.org/docs/latest/server_admin/#compromised-authorization-code)
+        *   _“For the [OIDC Auth Code Flow](https://www.keycloak.org/docs/latest/server_admin/#_oidc-auth-flows), it would be very hard for an attacker to compromise Keycloak authorization codes. Keycloak generates a cryptographically strong random value for its authorization codes so it would be very hard to guess an access token. An authorization code can only be used once to obtain an access token. In the admin console you can specify how long an authorization code is valid for on the [timeouts page](https://www.keycloak.org/docs/latest/server_admin/#_timeouts). This value should be really short, as short as a few seconds and just long enough for the client to make the request to obtain a token from the code._
+        *   _You can also mitigate against leaked authorization codes by applying PKCE to clients. See [Proof Key for Code Exchange (PKCE)](https://www.keycloak.org/docs/latest/server_admin/#_proof-key-for-code-exchange) to learn how.”_
+    *   _Eg. _Authorization Services and [handling custom polices implemented in JavaScript](https://www.keycloak.org/docs/latest/authorization_services/index.html#_policy_js)
+        *   _“By default, JavaScript Policies can not be uploaded to the server. You should prefer deploying your JS Policies directly to the server as described in [JavaScript Providers](https://www.keycloak.org/docs/latest/server_development/#_script_providers). If you still want to use the Keycloak Administration Console to manage your JS policies you should enable the ‘Upload Scripts’ feature.”_
+        *   This is done to mitigate any issues around handling (escaping/sandboxing) custom JavaScript code for those policies within the browser (within Admin UI).  
+*   User impersonation
+    *   Server provides ability for the administrator with a given set of permissions to [impersonate a particular user](https://www.keycloak.org/docs/latest/server_admin/#impersonation) and act on his behalf. Security of aspects of it can be configured on few different levels
```

Comment:
```suggestion
    *   Server provides ability for the administrator with a given set of permissions to [impersonate a particular user](https://www.keycloak.org/docs/latest/server_admin/#impersonation) and act on their behalf. Security of aspects of this can be configured on few different levels
```